client/pp-display/src/pp.tsx

```diff
@@ -262,82 +263,129 @@ const ppDisplay : FunctionComponent<PpProps> = (props) => {
         }
     };
 
-    const computeNeededBreaks = (maxBreaks: number) => {
+    const computeNeededBreaks = () => {
         if(container.current) {
-            if(content.current) {
-                const containerRect = container.current.getBoundingClientRect();
-                const contentRect = content.current.getBoundingClientRect();
-                if(containerRect.width < contentRect.width) {
-                    if(displayState.breakIds.length < maxBreaks) {
-                        checkBreaks(containerRect.width);
-                    }
-                }
+            const containerRect = container.current.getBoundingClientRect();
+            if(displayState.tokenStream && displayState.context) {
+                scanTokenStream(displayState.tokenStream, containerRect.width, displayState.context);
             }
         }
 
     };
 
-    const checkBreaks = (containerWidth: number) => {
-        if(content.current) {
-            let breakInfo : BreakInfo | null = null;
-            const boxes = content.current.querySelectorAll("."+classes.Box);
-            
-            for(let box of boxes) {
-
-                const parentBox = box.closest(`:not(#${box.id}).${classes.Box}`);
-                const parentOffset = parentBox ? parentBox.getBoundingClientRect().left : 0;
-                const offset =parentBox ? box.getBoundingClientRect().left - parentOffset : 0;
-
-                const breaks = box.querySelectorAll(`:scope > :not(.${classes.Box}):not(.${classes.Tag}):not(.${classes.Text})`);
-
-                for(let br of breaks) {
-                    const breakId = br.id;
-                    if(displayState.breakIds.find(info => info.id === breakId)) { continue; }
+    //create a canvas and get its context to compute character width
+    const getContext = ()  => {
+        const fragment = document.createDocumentFragment();
+        const canvas = document.createElement('canvas');
+        fragment.appendChild(canvas);
+        return canvas.getContext('2d');
+    };
 
-                    const next = br.nextElementSibling;
-                    if(next && next.getBoundingClientRect().right > containerWidth) {
-                        // const offset = offsetLevel;
-                        breakInfo = {id: breakId, offset: offset};
+    const estimateBoxWidth = (box: Box, context: CanvasRenderingContext2D) : number => {
+        let currentWidth = 0;
+        for (let childBox of box.boxChildren) {
+            if(childBox) {
+                switch(childBox.type) {
+                    case DisplayType.box:
+                        currentWidth += estimateBoxWidth(childBox, context);
+                        break;
+                    case DisplayType.break:
```

Comment:
I haven't tried out this, but I believe the estimation would be more accurate if
* a `Break` with `shouldBreak: true` resets the current width
* `estimateBoxWidth` returns the max width

Reply:
I am not sure what you mean ?
Currently the estimation of whether to break or not does not happen here but rather in the ```scanTokenStream``` function.
The box always has the max width but its only relevant to breaks if a box is a ```hvBox``` in which case all its breaks should happen.
Otherwise we use the estimation of a ```block``` which happens from break to break.

Reply:
Yeah, I agree there is definitely a more optimal way to do this. However ```buildTokenStream``` is only called once on receiving a ```ppString``` and in practice, even on examples that use to take a considerable amount of time, this version speeds everything up by an order of magnitude. I think there will be time later down the road to optimise this further.